File: 16/8-Two-Factor-Authentication/solution.exs
#!/usr/bin/env elixir

{width, height} = {50, 6}

pixels =
  "input.txt"
  |> File.read!()
  |> String.split("\n", trim: true)
  |> Enum.reduce(MapSet.new(), fn
    "rect " <> rest, acc ->
      [_, w, h] = Regex.run(~r/(\d+)x(\d+)/, rest)
      {w, h} = {String.to_integer(w), String.to_integer(h)}
      for x <- 0..(w - 1), y <- 0..(h - 1), into: acc, do: {x, y}

    "rotate row y=" <> rest, acc ->
      [_, r, amt] = Regex.run(~r/(\d+) by (\d+)/, rest)
      {r, amt} = {String.to_integer(r), String.to_integer(amt)}

      acc
      |> Enum.map(fn
        {x, ^r} -> {rem(x + amt, width), r}
        other -> other
      end)
      |> MapSet.new()

    "rotate column x=" <> rest, acc ->
      [_, c, amt] = Regex.run(~r/(\d+) by (\d+)/, rest)
      c = String.to_integer(c)
      amt = String.to_integer(amt)

      acc
      |> Enum.map(fn
        {^c, y} -> {c, rem(y + amt, height)}
        other -> other
      end)
      |> MapSet.new()
  end)

lit_count = MapSet.size(pixels)

IO.puts("Part 1 - Pixels lit: #{lit_count}\n")

IO.puts("Part 2 - Screen output:\n")

for y <- 0..(height - 1) do
  row =
    for x <- 0..(width - 1) do
      if MapSet.member?(pixels, {x, y}), do: "#", else: "."
    end
    |> Enum.join()

  IO.puts(row)
end

File: 16/8-Two-Factor-Authentication/solution.rb
#!/usr/bin/env ruby

# ABOUTME: Advent of Code 2016 Day 8 - Two-Factor Authentication solution
# ABOUTME: Simulates a screen with pixel operations (rect, rotate row, rotate column)

require 'set'

W, H = 50, 6

def parse(line)
  case line
  when /^rect (\d+)x(\d+)$/           then { t: :rect, w: $1.to_i, h: $2.to_i }
  when /^rotate row y=(\d+) by (\d+)$/ then { t: :row, r: $1.to_i, a: $2.to_i }
  when /^rotate column x=(\d+) by (\d+)$/ then { t: :col, c: $1.to_i, a: $2.to_i }
  end
end

def apply(pixels, i)
  case i[:t]
  when :rect then (0...i[:w]).to_a.product((0...i[:h]).to_a).each { pixels.add(_1) }; pixels
  when :row  then pixels.map { |x,y| y == i[:r] ? [(x + i[:a]) % W, y] : [x,y] }.to_set
  when :col  then pixels.map { |x,y| x == i[:c] ? [x, (y + i[:a]) % H] : [x,y] }.to_set
  end
end

pixels = File.readlines('input.txt', chomp: true).map(&method(:parse)).reduce(Set.new, &method(:apply))

puts "Part 1 - Pixels lit: #{pixels.size}\n\n"
puts "Part 2 - Screen output:\n\n"
H.times { |y| puts W.times.map { |x| pixels.include?([x,y]) ? '#' : '.' }.join }
